---
title: "Analysis on ERCC spike in"
author:
PROJECT_AUTHORS
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: ../references.bib
---

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
source("/etc/rstudio/Rprofile.site")
library(data.table)
require(dplyr)
require(DT)
library(ggplot2)
library(ggpubr)
library(pheatmap)
library(tidyr)
library(heatmaply)
library(plotly)
options(warn = -1)

Tag_this_analysis <- "ERCC_QC"

corr_figure <- function(ct, xlab, ylab) {
  yrange <- abs(max(c(ct[[xlab]], ct[[ylab]]))-min(c(ct[[xlab]], ct[[ylab]])))
  statcor_y <- max(c(ct[[xlab]], ct[[ylab]]))
  statregline_y <- max(c(ct[[xlab]], ct[[ylab]]))-0.05 * yrange
  fig <- ggplot(ct, aes(.data[[xlab]], .data[[ylab]])) +
         geom_point(size=1, color="royalblue", alpha=0.2) + theme_bw() +
         ggtitle(paste(xlab, "vs", ylab)) + theme(plot.title = element_text(hjust = 0.5)) +
         geom_smooth(method = "lm", color="gray", alpha=0.1) +
         xlab(xlab) + ylab(ylab) +
         stat_cor(label.y = statcor_y)+ #this means at 35th unit in the y axis, the r squared and p value will be shown
         stat_regline_equation(label.y = statregline_y) #this means at 30th unit regresion line equation will be shown
  fig
}

corr_figure_btw2samples <- function(title, sample1, sample2) {
  exptable <- fread("DGEA_All_samples_norm_counts.csv")
  columnsx <- c("gene_id", "gene_name", sample1)
  columnsy <- c("gene_id", "gene_name", sample2)
  xdata <- exptable[, ..columnsx]
  ydata <- exptable[, ..columnsy]
  
  merged_table <- merge(xdata, ydata,
                        by=c("gene_id","gene_name"), suffixes = c(".xlab", ".ylab"))
  yrange <- abs(max(ydata[[sample2]])-min(ydata[[sample2]]))
  statcor_y <- max(ydata[[sample2]])
  statregline_y <- max(ydata[[sample2]])-0.05 * yrange
  
  fig <- ggplot(merged_table, aes_string(sample1, sample2)) +
       geom_point(size=1, color="royalblue", alpha=0.2) + theme_bw() +
       ggtitle(title) + theme(plot.title = element_text(hjust = 0.5)) +
       geom_smooth(method = "lm", color="gray", alpha=0.1) +
       xlab(sample1) + ylab(sample2) +
       stat_cor(label.y = statcor_y)+ #this means at 35th unit in the y axis, the r squared and p value will be shown
       stat_regline_equation(label.y = statregline_y) #this means at 30th unit regresion line equation will be shown
  
   fig
}
# https://assets.thermofisher.com/TFS-Assets/LSG/manuals/cms_095046.txt
ercc_groups <- read.csv("thermofisher_LSG_manuals_cms_095046.txt", sep="\t")
ct <- fread(paste0(DIR_salmon, "salmon.merged.gene_tpm.tsv"))
ERCC_ct <- ct[startsWith(ct$gene_id, "ERCC"),]
ERCC_ct$gene_id <- gsub("_gene","",as.character(ERCC_ct$gene_id))
# ERCC_ct[,3:8] <- as.numeric(ERCC_ct[,3:8])
colnames(ERCC_ct) <- unlist(lapply(colnames(ERCC_ct), function(x) gsub("^X0","",x)))
```

### [Back to the main page](../Analysis_Report_RNAseq.html)

Variation in RNA expression data can be attributed to a variety of factors. In the Genomics Facility we committed to minimizing unwanted biases in quantitative RNA expression analysis. At the same time, we engaged a common set of external RNA controls  ERCC to control and evaluate the variability in RNA sequencing. The controls consist of a set of unlabeled, polyadenylated transcripts (92 transcripts, 250 to 2,000 nt in length) designed to be added to an RNA analysis experiment after sample isolation, in order to measure against defined performance criteria. For more details about the ERCC controls applied, please refer to [the documentation from the manufacture](https://www.thermofisher.com/order/catalog/product/4456739).

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=1.5, fig.height=1}
FC_tab <- merge(ERCC_ct, ercc_groups, by.x="gene_id", by.y="ERCC.ID")

plot_scatter_spikein <- function(FC_tab, whichmix, sample) {
  if (whichmix == "mix1") {
    mixcol <- "concentration.in.Mix.1..attomoles.ul."
  } else {
    mixcol <- "concentration.in.Mix.2..attomoles.ul."
  }
  sel_sample <- match(sample, colnames(FC_tab))
  sel_mix <- match(mixcol, colnames(FC_tab))
  sel_subgroupID <- match("subgroup", colnames(FC_tab))
  data <- FC_tab  %>%  dplyr::select(1,sel_subgroupID,sel_sample, sel_mix)
  data[,c(3)] <- log2(data[,c(3)]+1)
  data <- as.data.frame(data)
  colnames(data) <- c("gene_id","subgroup","norm_readcount","concentration")
  #data$concentration <- log10(data$concentration * 6.02 * 10^23+1)
  data$concentration <- log2(data$concentration)
  
  fit <- lm(norm_readcount ~ concentration, data = data)
  data$fit <- fitted(fit)
  r2 <- round(summary(fit)$adj.r.squared, digits =3)
  slope <- round(summary(fit)$coefficients[2,1],digits = 3)
  anno <- paste0("Slope = ",as.character(slope),"\n","R\u00B2 = ", as.character(r2))
  fig <- plot_ly(data = data, x = ~concentration, y = ~norm_readcount,  mode = "markers") %>%
         add_markers(y = ~norm_readcount, color = ~subgroup, colors = "Set1", text=~gene_id) %>% 
         add_trace(x = ~concentration, y = ~fit, mode = "lines", name="Regression", color="organge") %>%
        add_annotations(x= 10, y = 1,text= anno , showarrow = F )  %>%
         layout(title = paste(sample, whichmix),
                xaxis = list(zeroline = FALSE,
                             title=paste("Concentration in",whichmix,"(log2pM)")),
                yaxis = list(zeroline = FALSE,
                             title="TPM (log2)"))
  return(fig)
}

```

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE}
samples <- read.csv("../samplesheet.csv")
```

## Dynamic range and lower limit of detection {.tabset}

```{r,echo=FALSE, results='asis', warning=FALSE, message=FALSE}
for (sample in samples$sample) {
  cat("\n### ",  sample, "\n")
  fig <- plot_scatter_spikein(FC_tab, whichmix = "mix1", sample=sample)
  cat(htmltools::renderTags(as.widget(fig))$html)
  cat("\n")
}
```

<!-- ### SampleA -->

<!-- ```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=1.2, fig.height=1} -->
<!-- plot_scatter_spikein(FC_tab, whichmix = "mix1", sample="SampleA") -->
<!-- ``` -->

<!-- ## Ratios between Mix 1 and Mix 2 {.tabset} -->

<!-- ```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=1.5, fig.height=1} -->
<!-- # FC_tab[,3:11] <- FC_tab[,3:11]+1
# FC_tab$FC_3 <- FC_tab$Skull_3/FC_tab$Ulna_3
# FC_tab$FC_4 <- FC_tab$Skull_4/FC_tab$Ulna_4
# FC_tab$FC_5 <- FC_tab$Skull_5/FC_tab$Ulna_5
# FC_tab$FC_6 <- FC_tab$Skull_6/FC_tab$Ulna_6

# data_long <- FC_tab[,c(1,17, 18, 19, 20)] %>% pivot_longer(cols=c('FC_3', 'FC_4', 'FC_5', 'FC_6'),
#                     names_to='Group',
#                     values_to='FC')
# data_long$subgroup <- ercc_groups$subgroup[match(data_long$gene_id, ercc_groups$ERCC.ID)]
# data_long$expectedFC <- ercc_groups$expected.fold.change.ratio[match(data_long$gene_id, ercc_groups$ERCC.ID)]
# 
# write.table(FC_tab[,c("gene_id","subgroup",
#                       "concentration.in.Mix.1..attomoles.ul.","concentration.in.Mix.2..attomoles.ul.",
#                       "expected.fold.change.ratio","log2.Mix.1.Mix.2.",'FC_3', 'FC_4', 'FC_5', 'FC_6')],file = "ERCC_FC_table.csv") -->

<!-- plotratio <- function(data) { -->
<!--   fig <- plot_ly(data = data, x = ~expectedFC, y = ~FC, color = ~subgroup, colors = "Set1", symbol=~Group, text=~gene_id) -->

<!--   fig %>% -->
<!--     layout(shapes = list(list( -->
<!--       type = "line",  -->
<!--       x0 = 0,  -->
<!--       x1 = 4,  -->
<!--       xref = "x", -->
<!--       y0 = 0,  -->
<!--       y1 = 4,  -->
<!--       yref = "y", -->
<!--       line = list(color = "royalblue") -->
<!--     )), -->
<!--     yaxis = list(range = c(0,16))) -->

<!-- } -->
<!-- ``` -->


<!-- ```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=1.5, fig.height=1} -->
<!-- plotratio(data_long) -->
<!-- ``` -->


<!-- ### NEB11 -->

<!-- ```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=1.5, fig.height=1} -->
<!-- sel_data <- data_long[data_long$Group=="FC_NEB11",] -->
<!-- plotratio(sel_data) -->
<!-- ``` -->


## Heatmap

Heatmap for the overall spike-in ERCC expression detected in RNA sequencing.

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=1.5, fig.height=2}
heatmap_t <- ERCC_ct[, c(-1,-2)]
rownames(heatmap_t) <- ERCC_ct$gene_id
heatmap_t <- log10(heatmap_t+1)

heatmaply(heatmap_t, main = "Normalized read counts of spike-in",
          method = "plotly",labRow=ERCC_ct$gene_id,
          xlab = "Samples", ylab = "Spike-in",
          showticklabels = c(TRUE, FALSE), show_dendrogram = c(FALSE, TRUE),
          key.title = "Scaled\nexpression\nin log10 scale",
          label_names = c("Gene", "Sample", "Expression"))

```

<!-- According to the above heatmap, mix 1 (added to H889) and mix 2 (added to H891) are clustered properly. -->

<!-- # Correlation of H889 (Mix1) {.tabset} -->

<!-- ## NEB11 vs NEB12 -->

<!-- ```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=4, fig.height=4} -->
<!-- corr_figure(ct=heatmap_t, xlab="NEB11_H889", ylab="NEB12_H889") -->
<!-- ``` -->


<!-- # Correlation of H891 (Mix 2){.tabset} -->

<!-- ## NEB11 vs NEB12 -->

<!-- ```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=4, fig.height=4} -->
<!-- corr_figure(ct=heatmap_t, xlab="NEB11_H891", ylab="NEB12_H891") -->
<!-- ``` -->


# References

::: {#refs}
:::

# R session information

```{r, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```