---
title: "Differential Gene Expression Analysis for TITLEDESCRIPTION"
author:
GPM_AUTHORS
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
library("heatmaply")
require(dplyr)
require(DT)
library(ggplot2)
library(DESeq2)
library("tximport")
library("readr")
library("tximportData")
library(stringr)
library(tidyr)
library(data.table)
library('openxlsx')
library(AnnotationDbi, warn.conflicts = FALSE)
options(warn = -1)

DIR_base <- "GPM_DIR_BASE/"
DIR_project <- paste0(DIR_base, "analysis/")

source(paste0(DIR_project, "functions.R"))

load(paste0(DIR_project, "data.RData"))
load("SAMPLE_RData")

if (organism == "hsapiens") {
  library(org.Hs.eg.db)
} else if (organism == "mmusculus") {
  library(org.Mm.eg.db)
} else if (organism == "rnorvegicus") {
  library(org.Rn.eg.db)
}

Tag_this_analysis <- "FILETAG"
File_stats <- paste0('DGEA_',Tag_this_analysis,'_genes_stats.csv')
File_norm_expression <- paste0('DGEA_',Tag_this_analysis,'_norm_exp_quant.csv')

File_xlsx_res <- paste0('DGEA_',Tag_this_analysis,'_res.xlsx')
```

### [Back to front page](Analysis_Report_RNAseq.html)

Differential gene expression analysis is done with DESEQ2 package in R [@love2014moderated] and follows the instruction for Salmon quantification ([Analyzing RNA-seq data with DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)).

This analysis is done with PAIRED data.

```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
deseq_output <- run_deseq_salmon(filtered_samples, spikein=spikein_ERCC, paired=paired,
                                 countsFromAbundance=countsFromAbundance,
                                 lengthcorrection=lengthcorrection)
```

# Interactive Figures {.tabset}

## PCA

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
RNAseq_PCA_plotly(deseq_output$norm_count, filtered_samples)
```

## 3D-PCA

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
RNAseq_3D_PCA_plotly(deseq_output$norm_count, filtered_samples)
```

`r if(volcano){"## Volcano plot"}`

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, eval=volcano}
RNAseq_volcano_plotly(deseq_output$deseq2res)
```

`r if(maplot){"## MA plot"}`

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, eval=maplot}
RNAseq_maplot_plotly(deseq_output$deseq2res)
```

`r if(spikein_ERCC&maplot){"## MA plot of spike in"}`

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, eval=spikein_ERCC&maplot}
RNAseq_maplot_plotly_ERCC(deseq_output$res_combined_ERCC)
```

## Heatmap

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE}
RNAseq_heatmap_plotly(deseq_output$deseq2res)
```


# Static Figures {.tabset}

## PCA

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=6, fig.height=4}
RNAseq_PCA_ggplot2(deseq_output, filtered_samples)
```

`r if(volcano){"## Volcano plot"}`

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=6, fig.height=4, eval=volcano}
RNAseq_volcano_ggplot2(deseq_output)
```

`r if(maplot){"## MA plot"}`

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=6, fig.height=4, eval=maplot}
RNAseq_maplot_ggplot2(deseq_output)
```

## Heatmap

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=6, fig.height=4}
RNAseq_heatmap_ggplot2(deseq_output)
```

`r if(sigtable){"# Tables {.tabset}"}`

`r if(sigtable){"## Sig. genes"}`

```{r, layout="l-body-outset", echo=FALSE, results='asis', warning=FALSE, message=FALSE, eval=sigtable}
table_sig_genes(deseq_output$deseq2res_sig)
```

`r if(sigtable){"## Up genes"}`

```{r, layout="l-body-outset", echo=FALSE, results='asis', warning=FALSE, message=FALSE, eval=sigtable}
table_sig_genes(deseq_output$deseq2res_sig[deseq_output$deseq2res_sig$log2FoldChange > 0,])
```

`r if(sigtable){"## Down genes"}`

```{r, layout="l-body-outset", echo=FALSE, results='asis', warning=FALSE, message=FALSE, eval=sigtable}
table_sig_genes(deseq_output$deseq2res_sig[deseq_output$deseq2res_sig$log2FoldChange < 0,])
```

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
# download_table()
cols <- c("ENSEMBL", "SYMBOL", "GENENAME")
ensemblkeys <- gsub("\\.[0-9]*$", "", deseq_output$deseq2res$gene_id)
if (organism == "hsapiens") {
  genenames <- select(org.Hs.eg.db, keys=ensemblkeys, columns=cols, keytype="ENSEMBL")
} else if (organism == "mmusculus") {
  genenames <- select(org.Mm.eg.db, keys=ensemblkeys, columns=cols, keytype="ENSEMBL")
} else if (organism == "rnorvegicus"){
  genenames <- select(org.Rn.eg.db, keys=ensemblkeys, columns=cols, keytype="ENSEMBL")
}
genenames <- genenames[match(ensemblkeys, genenames$ENSEMBL), ]
deseq_output$deseq2res <- cbind(deseq_output$deseq2res, genenames)

write.csv(deseq_output$deseq2res,paste0(DIR_project, File_stats))
write.csv(deseq_output$norm_count,paste0(DIR_project, File_norm_expression))

wb <- createWorkbook()
addWorksheet(wb, "DGEA_stats")
writeData(wb, "DGEA_stats", deseq_output$deseq2res)
addWorksheet(wb, "norm_exp_quant")
writeData(wb, "norm_exp_quant", deseq_output$norm_count)
## Save workbook to working directory
saveWorkbook(wb, file = File_xlsx_res, overwrite = TRUE)
```

# Download statistics

* Download xlsx file for all results: [`r File_xlsx_res`](`r File_xlsx_res`) (However, CSV file is more recommended for robust analysis.)
* Download CSV file for the statistics of all valid genes in differential expression analysis (ENSEMBL ID, Gene Symbol and Gene name are included): [`r File_stats`](`r File_stats`)
* Download CSV file for the expression quantification of all genes: [`r File_norm_expression`](`r File_norm_expression`)

# References

<div id="refs"></div>


# R session information

```{r, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```
