---
title: "Simple Comparison for {{sample1}} and {{sample2}}"
author:
  - {{authors}}
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
    code_download: true
    code_folding: hide
bibliography: ../references.bib
---

```{r setup, echo=TRUE, warning=FALSE, message=FALSE}
source("/etc/rstudio/Rprofile.site")
# Load libraries
library(DESeq2)
library(dplyr)
library(ggplot2)
library(plotly)
library(viridis) 
library(DT)
library(tximport)
library(data.table)
library(openxlsx)
library(ggrepel)
library(tibble)
library(RColorBrewer)
library(pheatmap)

# Define samples to compare
sample1 <- {{sample1}}  # Define sample 1 name here
sample2 <- {{sample2}}    # Define sample 2 name here
organism <- "{{organism}}"  # e.g., "hsapiens", "mmusculus" PROJECT_PROCESSING_ORGANISM
######################## Below is optional #########################
# Define the base project path
project_base <- "PROJECT_PROJECT_PATH"
analysis_dir <- file.path(project_base, "analysis")
dgea_dir <- file.path(analysis_dir, "DGEA")
salmon_dir <- file.path(project_base, "PROJECT_PROCESSING_METHOD/results/star_salmon")
tx2gene_file <- file.path(salmon_dir, "tx2gene.tsv")

# Dynamically load the appropriate annotation package based on organism
if (organism == "mmusculus") {
  library(org.Mm.eg.db)
  annotation_db <- org.Mm.eg.db
} else if (organism == "hsapiens") {
  library(org.Hs.eg.db)
  annotation_db <- org.Hs.eg.db
} else {
  stop("Unsupported organism. Please specify 'mmusculus' or 'hsapiens'.")
}

####################################################################
#   Define run specifications
####################################################################
norm_spikein_ercc <- FALSE
highlighted_genes <- NA # e.g., c("Sub1", "", "")

cutoff_adj_p <- 0.05
cutoff_log2fc <- 1

# Loading samples (modify if necessary)
samplesheet <- read.csv("../../PROJECT_PROCESSING_METHOD/samplesheet.csv")
samplesheet <- samplesheet %>%
  dplyr::select(-2, -3, -4)
rownames(samplesheet) <- samplesheet$sample
```

### [Back to front page](../Analysis_Report_RNAseq.html)

```{r, echo=TRUE, warning=FALSE, message=FALSE}
if ("PROJECT_PROCESSING_METHOD" == "nfcore_RNAseq") {
  ct <- fread(paste0(salmon_dir, "salmon.merged.gene_counts_length_scaled.tsv"))
  }
if ("PROJECT_PROCESSING_METHOD" == "nfcore_3mRNAseq") {
  ct <- fread(paste0(salmon_dir, "salmon.merged.gene_counts_scaled.tsv"))
  }

# Dynamically add columns for log10-transformed values
ct[[paste0(sample1, "_log10")]] <- log10(ct[[sample1]] + 1)
ct[[paste0(sample2, "_log10")]] <- log10(ct[[sample2]] + 1)
```

## Scatter plot

```{r plotly-3d-pca, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
scatter_plot <- plot_ly(
  data = ct,
  x = ~get(paste0(sample1, "_log10")),
  y = ~get(paste0(sample2, "_log10")),
  text = ~paste("Gene:", gene_name),  # Add gene names as hover text
  mode = "markers",
  type = "scatter",
  marker = list(size = 5, color = 'royalblue', opacity = 0.7)
) %>%
  layout(
    title = paste("Scatter Plot of Gene Expression Between", sample1, "and", sample2),
    xaxis = list(title = paste(sample1, "Expression")),
    yaxis = list(title = paste(sample2, "Expression"))
  )

scatter_plot
```

## MA plot

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
# Avoid log of zero by adding a small constant
ct[[paste0(sample1, "_adjusted")]] <- ct[[sample1]] + 1
ct[[paste0(sample2, "_adjusted")]] <- ct[[sample2]] + 1

# Calculate M and A values
ct$M <- log2(ct[[paste0(sample2, "_adjusted")]] / ct[[paste0(sample1, "_adjusted")]])
ct$A <- (log2(ct[[paste0(sample2, "_adjusted")]]) + log2(ct[[paste0(sample1, "_adjusted")]])) / 2

# MA plot
ma_plot <- plot_ly(
  data = ct,
  x = ~A,
  y = ~M,
  text = ~paste("Gene:", gene_name),  # Add gene names as hover text
  mode = "markers",
  type = "scatter",
  marker = list(size = 5, color = 'royalblue', opacity = 0.7)
) %>%
  layout(
    title = paste("MA Plot of Gene Expression Between", sample1, "and", sample2),
    xaxis = list(title = "Mean Expression (A)"),
    yaxis = list(title = "Log2 Fold Change (M)")
  )

ma_plot
```

## Statistics

```{r statistics, layout="l-body-outset", echo=TRUE, results='asis', warning=FALSE, message=FALSE}
datatable(
  ct,
  extensions = c("FixedColumns", "Buttons"),
  filter = 'top',
  options = list(
    autoWidth = FALSE,
    dom = 'Blftip',
    pageLength = 10,
    searchHighlight = FALSE,
    buttons = c('copy', 'csv', 'excel', 'print'),
    scrollX = TRUE,
    fixedColumns = list(leftColumns = 3)
  ),
  class = c('compact cell-border stripe hover'),
  rownames = FALSE
) %>%
  formatRound(columns = c(3:10), digits = 2)
```

# Export

```{r export, echo=TRUE, results='markup', warning=FALSE, message=TRUE}
ct <- ct %>%
    mutate(
        gene_full_name = mapIds(
            annotation_db,
            keys = sub("\\..*", "", gene_id),  # Remove version numbers if present
            column = "GENENAME",
            keytype = "ENSEMBL",
            multiVals = "first"
        )
    ) %>%
    relocate(gene_full_name, .after = gene_name)

export_filename <- paste0("SimpleComparison_export.xlsx")
# Create a new workbook
wb <- createWorkbook()
# Add each data frame as a new sheet
addWorksheet(wb, "Statistics")
writeData(wb, "Statistics", ct, row.names = FALSE)
# Save the workbook
saveWorkbook(wb, export_filename, overwrite = TRUE)
export_csv <- paste0("SimpleComparison_stat.csv")
write.csv(ct, export_csv, row.names = FALSE)
```

- Please download the export excel file: [`r export_filename`](`r export_filename`)
- Or download the CSV file: [`r export_csv`](`r export_csv`)
- If you desire to annotate the volcano plot and annotate the genes you prefer, please try out our web app [AnnoFig](https://genomics.rwth-aachen.de/annofig/) with the CSV file above.

# References

<div id="refs"></div>

# R session information

```{r DGEA_FILETAG-session, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```