---
title: "PROJECT_PROJECT_STRING"
author:
RMD_AUTHORS
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r conig, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
source("/etc/rstudio/Rprofile.site")
library(knitr)
library(stringr)
library("readr")
library(tidyr)
library(dplyr)
library(openxlsx)
library(data.table)
library(yaml)
library(plotly)
library(reader)
options(warn = -1)
options(knitr.duplicate.label = "allow")
# File paths
DIR_base <- "PROJECT_PROJECT_PATH/"
DIR_analysis <- paste0(DIR_base, "analysis/")
DIR_result <- paste0(DIR_base, "nfcore_16S/results_16s")

source(paste0(DIR_analysis, "functions_16s.R"))

################# Parameters #####################################################
Fig_width <- 600
Fig_height <- 400



```
![](RMD_INSTITUTE_LOGO)

This document is the summary of the bioinformatic analysis done by Genomic Facility in IZKF, RWTH Aachen Uniklinik. Please contact us for any further discussion or collaboration. If these results are used in any poster or publications, please acknowledge our core facility with the text below:

> This work was supported by the Genomics Facility, a core facility of the Interdisciplinary Center for Clinical Research (IZKF) Aachen within the Faculty of Medicine at RWTH Aachen University.

Please refer to our [Publication Policy](https://genomics.rwth-aachen.de/services/publication-policy/).

***
# Download/Browse the files

1. [Raw Data](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_data): 
    * FASTQ files after demultiplexing
2. [Processed Data](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data): 
    * results generated during the analsis process
3. [Reports](EXPORT_URL/PROJECT_PROJECT_NAME/3_Reports):
    * [MultiQC reports](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data/results/multiqc/multiqc_report.html) for the quality check
    * [nf-core summary report](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data/results/summary_report/summary_report.html)


You can also download everything by the commands below with the given username and password from us:

```
wget -r -nH -np --cut-dirs=2 -l 8 -P PROJECT_PROJECT_NAME --user=PROJECT_EXPORT_USER --password=PROJECT_EXPORT_PASSWORD --reject "index.html?*,index.html.tmp*,robots.txt*,robots.txt.tmp" https://genomics.rwth-aachen.de/data/PROJECT_PROJECT_NAME/
```

Please make sure that you store these data in a lab-wise level. Please refer to our [Data Management Policy](https://genomics.rwth-aachen.de/services/data-management-policy/).

\textcolor{red}{Your data will be deleted according to this policy automatically. Please download everything.}

# Method description

FASTQ files were generated using bcl2fastq (Illumina). To facilitate reproducible analysis, samples were processed using the publicly available nf-core/ampliseq pipeline version 2.12.0 [@Straub_Blackwell_Langarica-Fuentes_Peltzer_Nahnsen_Kleindienst_2020]. The pipeline was executed with Nextflow v24.10.3 [@di2017nextflow] using Docker 20.10.12 [@merkel2014docker] with the minimal command. In breif, the sequencing reads are quality controlled with FastQC and then trimmed with Cutadapt. Amplicon sequence variants were inferred with DADA2. Taxonomic classification were achieved with DADA2 and QIIME2. Absolute and relative feature/taxa count tables and plots, plots alpha rarefaction curves, computes alpha and beta diversity indices and plots were produced with QIIME2. SILVA was applied as reference taxomomy.

<!-- inserting points -->

# References

<div id="refs"></div>


# R session information

```{r info, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```