---
title: "PROJECT_PROJECT_STRING"
author: 
RMD_AUTHORS
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r conig, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
source("/etc/rstudio/Rprofile.site")
library("heatmaply")
library("readr")
library(tidyr)
library(dplyr)
library(openxlsx)
library(kableExtra)
library(data.table)
options(warn = -1)

################# File paths #####################################################
DIR_base <- "PROJECT_PROJECT_PATH/"
DIR_project <- paste0(DIR_base, "analysis/")

source(paste0(DIR_project, "report_functions.R"))
```


![](RMD_INSTITUTE_LOGO)

This document is the summary of the bioinformatic analysis done by Genomic Facility in IZKF, RWTH Aachen Uniklinik. Please contact us for any further discussion or collaboration. If these results are used in any poster or publications, please acknowledge our core facility with the text below:

> This work was supported by the Genomics Facility, a core facility of the Interdisciplinary Center for Clinical Research (IZKF) Aachen within the Faculty of Medicine at RWTH Aachen University.

Please refer to our [Publication Policy](https://genomics.rwth-aachen.de/services/publication-policy/).

***

In this report, you are able to:

* Browse and download the files and reports in an intuitive folder structure.
* Read the QC reports or bioinformatics reports online.
* Explore the results by the interactive figures in this report. You are able to search, select, zoom in or export the image.
* Export any tables in this report for gene lists or statistics in various format you want.
* Get the summary of the method description and citation.

After delivering this report, we can schedule a short meeting to discuss the result, otherwise, we will close this project with this basic analysis. If you have further request and questions, please don't hesitate to contact us. 

***

# Download/Browse the files

Please click the hyperlinked list below:

1. [Raw Data](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_data): 
    * FASTQ files after demultiplexing
    * [MultiQC report for demultiplexing](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_data/FASTQ_multiQC)
2. [Processed Data](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data): 
    * FastQC - read quality control
    * UMI-tools extract - UMI barcode extraction
    * UMI-collapse deduplicate - read deduplication
    * FastP - adapter trimming
    * Bowtie2 - contamination filtering
    * Bowtie - alignment against mature miRNAs and miRNA precursors (hairpins)
    * SAMtools - alignment result processing and feature counting
    * edgeR - normalization, MDS plot and sample pairwise distance heatmap
    * Bowtie - alignment against reference genome for QC purpose
    * mirtop - miRNA and isomiR annotation
    * miRDeep2 - known and novel miRNA annotation
    * miRTrace - a comprehensive tool for QC purpose
    * MultiQC - aggregate report, describing results of the whole pipeline
    * Pipeline information - Report metrics generated during the workflow execution
3. [Reports](EXPORT_URL/PROJECT_PROJECT_NAME/3_Reports):
    * [MultiQC reports](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data/nfcore_miRNAseq/results/multiqc/multiqc_report.html) for the quality check
    * Basic analysis of the Genomics Facility (This report)
  
In the command line environment, you can also download everything by the commands below with the given username and password from us:

```
wget -r -nH -np -P PROJECT_PROJECT_NAME --user=PROJECT_EXPORT_USER --password=PROJECT_EXPORT_PASSWORD --reject "index.html?*,index.html.tmp*,robots.txt*,robots.txt.tmp" https://genomics.rwth-aachen.de/data/PROJECT_PROJECT_NAME/
```

Please make sure that you store these data in a lab-wise level. Please refer to our [Data Management Policy](https://genomics.rwth-aachen.de/services/data-management-policy/).


# Differential Expression Analysis

## [Tumor vs Healthy](DEAmiRNA_Tumor_vs_Healthy.html)

# Sample Overview

```{r echo=FALSE, results='asis'}
samples <- read.csv("samplesheet.csv")
rownames(samples) <- samples$sample
kable(samples) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE)
```

# References

<div id="refs"></div>


# R session information

```{r info, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```