---
title: "PROJECT_PROJECT_STRING"
author: 
PROJECT_AUTHORS
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r init, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
source("/etc/rstudio/Rprofile.site")
library(dplyr)
library(kableExtra)
options(warn = -1)
options(knitr.duplicate.label = "allow")
# File paths
DIR_base <- "PROJECT_PROJECT_PATH/"
DIR_analysis <- paste0(DIR_base, "analysis/")

source(paste0(DIR_analysis, "report_functions.R"))
```

![](RMD_INSTITUTE_LOGO)

This document is the summary of the bioinformatic analysis done by Genomic Facility in IZKF, RWTH Aachen Uniklinik. Please contact us for any further discussion or collaboration. If these results are used in any poster or publications, please acknowledge our core facility with the text below:

> This work was supported by the Genomics Facility, a core facility of the Interdisciplinary Center for Clinical Research (IZKF) Aachen within the Faculty of Medicine at RWTH Aachen University.

Please refer to our [Publication Policy](https://genomics.rwth-aachen.de/services/publication-policy/).

***

# Download/Browse the files

1. [Raw Data](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_data): 
    * FASTQ files after demultiplexing
2. [Processed Data](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data): 
    * BAM files after alignment
    * BigWig files after normalization
    * Quantification tables from the counts of reads or quantification of the expressions
3. [Reports](EXPORT_URL/PROJECT_PROJECT_NAME/3_Reports):
    * [MultiQC reports](EXPORT_URL/PROJECT_PROJECT_NAME/3_Reports/multiqc/multiqc_report.html) for the quality check
    * Basic analysis of the Genomics Facility (This report)
    <!-- * [QC with ERCC spike-ins](ERCC/ERCC.html) -->
  
You can also download everything by the commands below with the given username and password from us:

```
wget -r -nH -np --cut-dirs=2 -l 8 -P PROJECT_PROJECT_NAME --user=PROJECT_EXPORT_USER --password=PROJECT_EXPORT_PASSWORD --reject "index.html?*,index.html.tmp*,robots.txt*,robots.txt.tmp" https://genomics.rwth-aachen.de/data/PROJECT_PROJECT_NAME/
```

Please make sure that you store these data in a lab-wise level. Please refer to our [Data Management Policy](https://genomics.rwth-aachen.de/services/data-management-policy/).

\textcolor{red}{Your data will be deleted according to this policy automatically. Please download everything.}

# Differential Gene Expression Analysis

In 'Group2 vs Group1' analysis, we compare gene expression between Group2 and Group1. 'Up' genes are expressed more in Group2 than Group1, indicating potential biological changes. 'Down' genes show lower expression in Group2, suggesting differences in gene activity.

### [All samples](DGEA/DGEA_All_samples.html)
<!-- INSERTING POINT -->

# References

<div id="refs"></div>


# R session information

```{r info, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```