---
title: "PROJECT_PROJECT_STRING"
author:
PROJECT_AUTHORS
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r conig, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
source("/etc/rstudio/Rprofile.site")
library(stringr)
library("readr")
# library(tidyr)
# library(data.table)
options(warn = -1)
options(knitr.duplicate.label = "allow")

################# File paths #####################################################
DIR_base <- "PROJECT_PROJECT_PATH/"
DIR_project <- paste0(DIR_base, "analysis/")
DIR_cellranger <- paste0(DIR_base, "cellranger_cellplex/")
DIR_multi_output <- paste0(DIR_cellranger, "results/")
DIR_multi_output_published <- "EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data/cellranger_cellplex/results/"
```

![](RMD_INSTITUTE_LOGO)

This document is the summary of the bioinformatic analysis done by Genomic Facility in IZKF, RWTH Aachen Uniklinik. Please contact us for any further discussion or collaboration. If these results are used in any poster or publications, please acknowledge our core facility with the text below:

> This work was supported by the Genomics Facility, a core facility of the Interdisciplinary Center for Clinical Research (IZKF) Aachen within the Faculty of Medicine at RWTH Aachen University.

Please refer to our [Publication Policy](https://genomics.rwth-aachen.de/services/publication-policy/).

***

# Download/Browse the files

1. [Raw Data](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_data): 
    * FASTQ files after demultiplexing
2. [Processed Data](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data): 
    * multi output for count and multiplexing_analysis. 
      + bam and bai for unassigned alignments
      + raw feature bc count matrix
      + tag calls
      
    * per sample result: 
      + summary for the results
      + cloupe file for browsing in Loupe browser
      + BAM and bai files after alignment
      + filtered feature bc count matrix
      + further analysis results
      
3. [Reports](EXPORT_URL/PROJECT_PROJECT_NAME/3_Reports):
    * [MultiQC reports](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_data/FASTQ_multiQC) for the quality check
    * Basic analysis of the Genomics Facility (This report)
  
In the command line environment, you can also download everything by the commands below with the given username and password from us:

```
wget -r -nH -np --cut-dirs=2 -l 8 -P PROJECT_PROJECT_NAME --user=PROJECT_EXPORT_USER --password=PROJECT_EXPORT_PASSWORD --reject "index.html?*,index.html.tmp*,robots.txt*,robots.txt.tmp" https://genomics.rwth-aachen.de/data/PROJECT_PROJECT_NAME/
```

Please make sure that you store these data in a lab-wise level. Please refer to our [Data Management Policy](https://genomics.rwth-aachen.de/services/data-management-policy/).

\textcolor{red}{Your data will be deleted according to this policy automatically. Please download everything.}

# Method description

FASTQ files from both gene expression and CellPlex libraries were generated using cellranger mkfastq (v7.1.0 10x Genomics). The generated fastq files were then analyzed by the cellranger multi (v7.1.0 10x Genomics) using the CMO Feature Barcode information. This report was generated using custom scripts in R version 4.2.2.


```{r load, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
samples <- list.dirs(path = DIR_multi_output, full.names = F, recursive = F)
summaryFiles <- vector(mode="list", length=length(samples))
names(summaryFiles) <- samples

for (i in 1:length(samples)) {
    
        cells <- list.dirs(path = paste0(DIR_multi_output, samples[i],"/outs/per_sample_outs"), full.names = F, recursive = F)
        summaryFiles[[i]] <- vector(mode="list", length=length(cells))
    
    for ( j in 1:length(cells)) {
            summaryFiles[[i]][[j]] <- paste0(DIR_multi_output_published,samples[[i]],"/outs/per_sample_outs/",cells[[j]]",/web_summary.html")
    }
  
}
```

```{r, echo=FALSE, results="asis"}
for( i in 1:length(samples)){
    cat("\n")
    cat("# Summary reports of Cellranger multi for ", samples[[i]],sep="") #list sample first
    cat("\n")
    for (j in 1:length(cells)) {
        cat("\n")
        cat("## ", "[",cells[[j]],"]","(",summaryFiles[[i]][[j]],")\n", sep="") # Create links to the
        cat("\n")

    }
  
}
```

<!-- inserting points -->

# References

<div id="refs"></div>

# R session information

```{r info, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```