---
title: "PROJECT_PROJECT_STRING"
author: 
RMD_AUTHORS
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r init, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
options(warn = -1)
options(knitr.duplicate.label = "allow")
# File paths
DIR_base <- "PROJECT_PROJECT_PATH/"
DIR_analysis <- paste0(DIR_base, "analysis/")

source(paste0(DIR_analysis, "report_functions.R"))

generate_Rmd <- TRUE
render_html <- TRUE
```

![](RMD_INSTITUTE_LOGO)

This document is the summary of the bioinformatic analysis done by Genomic Facility in IZKF, RWTH Aachen Uniklinik. Please contact us for any further discussion or collaboration. If these results are used in any poster or publications, please acknowledge our core facility with the text below:

    > This work was supported by the Genomics Facility, a core facility of the Interdisciplinary Center for Clinical Research (IZKF) Aachen within the Faculty of Medicine at RWTH Aachen University.

    Please refer to our [Publication Policy](https://genomics.rwth-aachen.de/services/publication-policy/).

    ***

# Download/Browse the files

1. [Raw Data](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_Data): 
    * BCL files from the sequencer
    * FASTQ files after demultiplexing
2. [Processed Data](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data): 
    * BAM files after alignment
    * BigWig files after normalization
    * Quantification tables from the counts of reads or quantification of the expressions
3. [Reports](EXPORT_URL/PROJECT_PROJECT_NAME/3_Reports):
    * [MultiQC reports](EXPORT_URL/PROJECT_PROJECT_NAME/3_Reports/multiqc/multiqc_report.html) for the quality check
    * Basic analysis of the Genomics Facility (This report)
4. [Compressed files for download](EXPORT_URL/PROJECT_PROJECT_NAME/compressed_tar/):
    * Three tar files are available for download the above three folders.
  
You can also download everything by the commands below with the given username and password from us:

```
wget -r -nH -np -nd -P --user USERNAME --password PASSWORD -A '*.tar' EXPORT_URL/PROJECT_PROJECT_NAME/compressed_tar/ -P PROJECT_PROJECT_NAME 
```

Please make sure that you store these data in a lab-wise level. Please refer to our [Data Management Policy](https://genomics.rwth-aachen.de/services/data-management-policy/).

\textcolor{red}{Your data will be deleted according to this policy automatically. Please download everything.}

# Sample Overview

```{r echo=FALSE, results='asis'}
samplesheet_file <- read.csv("samplesheet.csv")
# samplesheet_file <- subset(samplesheet_file, select = -c(fastq_1, fastq_2))
kable(samplesheet_file) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE)
```

# Differential Gene Expression Analysis

```{r init, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
DIR_DGEA <- paste0(DIR_analysis, "DGEA/")
source(paste0(DIR_DGEA, "DGEA_RNAseq_functions.R"))
spikein_ERCC <- FALSE
paired_test <- FALSE
# 3mRNAseq
countsFromAbundance <- "no"
lengthcorrection <- FALSE
# RNAseq, totalRNAseq, mRNAseq
countsFromAbundance <- "lengthScaledTPM"
lengthcorrection <- TRUE

organism <- "" # hsapiens mmusculus rnorvegicus sscrofa
organism_DB <- organism_DB(organism)

```

### All Samples 3d PCA plot

```{r, echo=FALSE, results="asis", warning=FALSE, message=FALSE, dpi=600, fig.width=6, fig.height=4}
description <- "All samples"
filtered_samples <- samples[complete.cases(samples$group),]
deseq_output <- run_deseq_salmon(filtered_samples, 
                                 spikein=spikein_ERCC, 
                                 paired=paired_test,
                                 countsFromAbundance=countsFromAbundance,
                                 lengthcorrection=lengthcorrection)
RNAseq_3D_PCA_plotly(deseq_output$norm_count, filtered_samples)
```

```{r, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
Tag_this_analysis <- "All_Samples"
File_stats <- paste0('DGEA_',Tag_this_analysis,'_genes_stats.csv')
File_norm_expression <- paste0('DGEA_',Tag_this_analysis,'_norm_exp_quant.csv')
File_xlsx_res <- paste0('DGEA_',Tag_this_analysis,'_res.xlsx')

cols <- c("ENSEMBL", "SYMBOL", "GENENAME")
ensemblkeys <- gsub("\\.[0-9]*$", "", deseq_output$deseq2res$gene_id)
genenames <- select(organsim_DB, keys=ensemblkeys, columns=cols, keytype="ENSEMBL")
genenames <- genenames[match(ensemblkeys, genenames$ENSEMBL), ]
deseq_output$deseq2res <- cbind(deseq_output$deseq2res, genenames)
write.csv(deseq_output$deseq2res, paste0(DIR_DGEA, File_stats))
write.csv(deseq_output$norm_count, paste0(DIR_DGEA, File_norm_expression))

wb <- createWorkbook()
addWorksheet(wb, "DGEA_stats")
writeData(wb, "DGEA_stats", deseq_output$deseq2res)
addWorksheet(wb, "norm_exp_quant")
writeData(wb, "norm_exp_quant", deseq_output$norm_count)
saveWorkbook(wb, file = paste0(DIR_DGEA, File_xlsx_res), overwrite = TRUE)
```

* Download CSV file for the expression quantification of all 'All Samples' genes: [`r File_norm_expression`](`r paste0("DGEA/",File_norm_expression)`)

<!-- ### [Correlation analysis](DGEA_correlation_analysis.html) -->


```{r, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
filetag <- "Control_vs_Treatment"
description <- str_replace_all(filetag, "_", " ")
filtered_samples <- samples
filtered_samples$group <- factor(filtered_samples$group, levels = c("Control","Treatment"))
if ( generate_Rmd == TRUE ) {
  add_DGEA(description, filetag, filtered_samples, paired=paired_test)
}
if ( render_html == TRUE ) {
  rmarkdown::render(paste0(DIR_DGEA, 'DGEA_',filetag,'.Rmd'), 
                    output_format = 'html_document',
                    output_file = paste0(DIR_DGEA,'DGEA_',filetag,'.html'))
}

```
### [`r description`](`r paste0(DIR_DGEA, 'DGEA_',filetag,'.html')`)

<!-- INSERTING POINT -->

# References

<div id="refs"></div>


# R session information

```{r info, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```