---
title: "PROJECT_PROJECT_STRING"
author: 
RMD_AUTHORS
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r init, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
library(dplyr)
library(kableExtra)
options(warn = -1)
options(knitr.duplicate.label = "allow")
# File paths
DIR_base <- "PROJECT_PROJECT_PATH/"
DIR_analysis <- paste0(DIR_base, "analysis/")

source(paste0(DIR_analysis, "report_functions.R"))
```

![](RMD_INSTITUTE_LOGO)

This document is the summary of the bioinformatic analysis done by Genomic Facility in IZKF, RWTH Aachen Uniklinik. Please contact us for any further discussion or collaboration. If these results are used in any poster or publications, please acknowledge our core facility with the text below:

> This work was supported by the Genomics Facility, a core facility of the Interdisciplinary Center for Clinical Research (IZKF) Aachen within the Faculty of Medicine at RWTH Aachen University.

Please refer to our [Publication Policy](https://genomics.rwth-aachen.de/services/publication-policy/).

***

# Download/Browse the files

1. [Raw Data](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_data): 
    * BCL files from the sequencer
    * FASTQ files after demultiplexing
    * [MultiQC of Demultiplexing](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_data/FASTQ/multiqc/multiqc_report.html)
2. [Processed Data](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data): 
   Including all the configuration and outputs from [nf-core/methylseq](https://nf-co.re/methylseq), including the steps below:
    * FastQC - Raw read QC
    * TrimGalore - Adapter trimming
    * Alignment - Aligning reads to reference genome
    * Deduplication - Deduplicating reads
    * Methylation Extraction - Calling cytosine methylation steps
    * [Bismark Reports](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data/nfcore_methylseq/results/bismark/reports/) - Single-sample and [summary analysis reports](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data/nfcore_methylseq/results/bismark/summary/bismark_summary_report.html)
    * [Qualimap](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data/nfcore_methylseq/results/qualimap/) - Tool for genome alignments QC
    * Preseq - Tool for estimating sample complexity
    * [MultiQC](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data/nfcore_methylseq/results/multiqc/bismark/multiqc_report.html) - Aggregate report describing results and QC from the whole pipeline
    * Pipeline information - Report metrics generated during the workflow execution
3. [Reports](EXPORT_URL/PROJECT_PROJECT_NAME/3_Reports):
    * Basic analysis of the Genomics Facility (This report)
4. [Compressed files for download](EXPORT_URL/PROJECT_PROJECT_NAME/compressed_tar/):
    * Three tar files are available for download the above three folders.
  
You can also download everything by the commands below with the given username and password from us:

```
wget -r -nH -np -nd --user PROJECT_EXPORT_USER --password PROJECT_EXPORT_PASSWORD -A '*.tar*' EXPORT_URL/PROJECT_PROJECT_NAME/compressed_tar/ -P PROJECT_PROJECT_NAME
```

Please make sure that you store these data in a lab-wise level. Please refer to our [Data Management Policy](https://genomics.rwth-aachen.de/services/data-management-policy/).

\textcolor{red}{Your data will be deleted according to this policy automatically. Please download everything.}

# Method description

We employed the nfcore/methylseq [@phil_ewels_2024_10463781] pipeline for comprehensive analysis of DNA methylation patterns, crucial for understanding epigenetic regulation and its implications in various biological processes and diseases. Following quality control using FastQC and Trimmomatic, clean reads were aligned to the reference genome using Bismark to account for bisulfite conversion and identify methylated cytosines.

# Sample Overview

```{r echo=FALSE, results='asis'}
samples <- read.csv("samplesheet.csv")
rownames(samples) <- samples$sample
kable(samples) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = FALSE)
```

<!-- INSERTING POINT -->

# References

<div id="refs"></div>


# R session information

```{r info, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```