---
title: "PROJECT_PROJECT_STRING"
author: 
RMD_AUTHORS
date: "Last compiled on `r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: false
    theme: lumen
bibliography: references.bib
---

```{r conig, echo=FALSE, results="hide", warning=FALSE, message=FALSE}
source("/etc/rstudio/Rprofile.site")
library("heatmaply")
library("readr")
options(warn = -1)

DIR_base <- "PROJECT_PROJECT_PATH/"
DIR_project <- paste0(DIR_base, "analysis/")
DIR_nfcore <- paste0(DIR_base, "nfcore_ChIPseq/")
DIR_nfcore_res <- paste0(DIR_base, "nfcore_ChIPseq/results/")

source(paste0(DIR_project, "report_functions.R"))

## Load sample sheets from nf-core configuration
samples <- read.table(file.path(DIR_nfcore,"samplesheet.csv"), header=TRUE, sep = ",")
rownames(samples) <- samples$sample
```


![](RMD_INSTITUTE_LOGO)

This document is the summary of the bioinformatic analysis done by Genomic Facility in IZKF, RWTH Aachen Uniklinik. Please contact us for any further discussion or collaboration. If these results are used in any poster or publications, please acknowledge our core facility with the text below:

> This work was supported by the Genomics Facility, a core facility of the Interdisciplinary Center for Clinical Research (IZKF) Aachen within the Faculty of Medicine at RWTH Aachen University.

Please refer to our [Publication Policy](https://genomics.rwth-aachen.de/services/publication-policy/).


***

# Download/Browse the files

Please click the hyperlinked list below:

1. [Raw Data](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_data): 
    * FASTQ files after demultiplexing
    * [MultiQC for FASTQ files](EXPORT_URL/PROJECT_PROJECT_NAME/1_Raw_data/FASTQ_multiQC)
2. [Processed Data](EXPORT_URL/PROJECT_PROJECT_NAME/2_Processed_data): 
    * BAM files after alignment
    * BigWig files after normalization
    * BED files for peaks by MACS2
    * The complete output from nfcore/chipseq pipeline
3. [Reports](EXPORT_URL/PROJECT_PROJECT_NAME/3_Reports):
    * [MultiQC reports](EXPORT_URL/PROJECT_PROJECT_NAME/3_Reports/multiqc/multiqc_report.html) for the quality check
    * Basic analysis of the Genomics Facility (This report)
  
In the command line environment, you can also download everything by the commands below with the given username and password from us:

```
wget -r -nH -np -P PROJECT_PROJECT_NAME --user=PROJECT_EXPORT_USER --password=PROJECT_EXPORT_PASSWORD --reject "index.html?*,index.html.tmp*,robots.txt*,robots.txt.tmp" https://genomics.rwth-aachen.de/data/PROJECT_PROJECT_NAME/
```

Please make sure that you store these data in a lab-wise level. Please refer to our [Data Management Policy](https://genomics.rwth-aachen.de/services/data-management-policy/).

# Method description

FASTQ files were generated using bcl2fastq (Illumina). To facilitate reproducible analysis, samples were processed using the publicly available [nf-core/ChIP-Seq pipeline](https://nf-co.re/chipseq/) version 2.0.0 [@ewels2020nf] implemented in Nextflow 23.10.1 [@di2017nextflow] using Docker 24.0.7 [@merkel2014docker] with the minimal command. In brief, lane-level reads were trimmed using Trim Galore 0.6.7 [@felix_krueger_2021_5127899] and aligned to the human genome (GRCh38.v44) using BWA 0.7.17 [@li2010fast]. Peaking calling was performed by MACS2 2.2.7 [@zhang2008model]. All analysis was performed using custom scripts in R version 4.2.2. 


# References

<div id="refs"></div>


# R session information

```{r info, echo=FALSE, results='markup', warning=FALSE, message=TRUE}
sessionInfo()
```