{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {
    "vscode": {
     "languageId": "raw"
    }
   },
   "source": [
    "---\n",
    "title: \"QC of scRNA-seq\"\n",
    "author:\n",
    "PROJECT_AUTHORS\n",
    "date: today\n",
    "format:\n",
    "  html:\n",
    "    toc: true\n",
    "    toc-depth: 3\n",
    "    html-math-method: katex\n",
    "    code-fold: true\n",
    "    embed-resources: true\n",
    "    page-layout: full\n",
    "execute:\n",
    "    echo: true\n",
    "    warning: false\n",
    "jupyter:\n",
    "  kernelspec:\n",
    "    name: \"pixi-kernel-python3\"\n",
    "    language: \"python\"\n",
    "    display_name: \"pixi-kernel-python3\"\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: true\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "from numbers import Number\n",
    "\n",
    "\"in the following cell, override the default pipeline parameters if needed\"\n",
    "\n",
    "# QC Params\n",
    "\n",
    "# check QC and filter cells for each sample separately, default: True.\n",
    "QC_PER_SAMPLE: bool = True\n",
    "\n",
    "# Correct ambient RNA, uses DecontX.\n",
    "CORRECT_AMBIENT_RNA: bool = False\n",
    "\n",
    "# Filter doublets using Scrublet\n",
    "FILTER_DOUBLETS: bool = False\n",
    "\n",
    "# Calculate cell cycle scores, based on scanpy implementation.\n",
    "CELL_CYCLE_SCORE: bool = True\n",
    "\n",
    "\n",
    "############################################################\n",
    "####################     QC dict      ######################\n",
    "############################################################\n",
    "qc_dict: dict[str, list] | dict[str, Number] = {\n",
    "    \"n_genes_by_counts\": 5,\n",
    "    \"total_counts\": 5,\n",
    "    \"pct_counts_mt\": 3,\n",
    "}  # type: ignore\n",
    "\n",
    "# MODIFY_ME, Specifiy only after running diagnotics on the samples\n",
    "# global config\n",
    "# qc_dict: dict[str, list] | dict[str, list] = {\n",
    "#     \"pct_counts_mt\": [0, 10],\n",
    "#     \"pct_counts_rb\": [0, 10],\n",
    "#     \"n_genes_by_counts\": [1500, 9000],\n",
    "#     \"total_counts\": [3_000, 40_000]\n",
    "# }\n",
    "\n",
    "\n",
    "# # per-sample config\n",
    "# qc_dict: dict[str, list] | dict[str, list] = {\n",
    "#     \"sample1\": {\n",
    "#         \"pct_counts_mt\": [0, 10],\n",
    "#         \"pct_counts_rb\": 5,\n",
    "#         \"total_counts\": [3_000, 40_000]\n",
    "#     },\n",
    "#     \"sample2\": {\n",
    "#         \"pct_counts_rb\": 3,\n",
    "#         \"n_genes_by_counts\": [1500, 9000],\n",
    "#         \"total_counts\": [3_000, 40_000]\n",
    "#     },\n",
    "#     \"sample3\": {\n",
    "#         \"pct_counts_mt\": [0, 10],\n",
    "#         \"n_genes_by_counts\": [1500, 9000],\n",
    "#         \"total_counts\": [3_000, 40_000]\n",
    "#     },\n",
    "#     \"sample4\": {\n",
    "#         \"pct_counts_mt\": 4\n",
    "#         \"pct_counts_rb\": 5,\n",
    "#         \"n_genes_by_counts\": 6,\n",
    "#     }\n",
    "# }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "import logging\n",
    "from collections.abc import Callable\n",
    "from functools import reduce\n",
    "from numbers import Number\n",
    "from operator import add\n",
    "from os import cpu_count, listdir, mkdir, path, walk\n",
    "\n",
    "import anndata as ad\n",
    "import hvplot.pandas\n",
    "import itables\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import panel as pn\n",
    "import scanpy as sc\n",
    "import seaborn as sns\n",
    "import session_info\n",
    "import tomlkit\n",
    "import utils.preprocessing_funcs as pf\n",
    "from matplotlib import rc_context\n",
    "from utils.autodiscover import AutoDiscover, discover_factory\n",
    "\n",
    "logging.basicConfig(level=logging.ERROR)\n",
    "sc.set_figure_params(dpi=100, dpi_save=300, format=\"png\")\n",
    "sc.settings.n_jobs = int(cpu_count() * 0.8)\n",
    "pn.extension()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "# Pipeline parameters\n",
    "with open(\"../config.toml\", \"r\") as f:\n",
    "    config = tomlkit.parse(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "\n",
    "# Directories\n",
    "ROOT_DIR = config[\"basic\"][\"ANALYSIS_DIR\"]\n",
    "DIR_SAVE = path.join(ROOT_DIR, config[\"basic\"][\"DIR_SAVE\"])\n",
    "DIR_samples = config[\"basic\"][\"DIR_SAMPLES\"]\n",
    "DIR_RESOURCES = config[\"basic\"][\"RESOURCE_PATH\"]\n",
    "\n",
    "# Basic information\n",
    "TECHNOLOGY: str = config[\"basic\"][\"TECHNOLOGY\"]\n",
    "ORGANISM: str = config[\"basic\"][\"ORGANISM\"]\n",
    "AUTODISCOVER: bool = config[\"basic\"][\"auto_find\"]\n",
    "samples: dict[str, str] = config[\"basic\"][\"samples\"]\n",
    "raw_samples: dict[str, str] = config[\"basic\"][\"raw_samples\"]\n",
    "\n",
    "# MAX_ONLY: ignores the lower bound.\n",
    "# LOG_TRANSFORMED: applies log1p transformation for the given columns.\n",
    "MAX_ONLY: list[str] = [\"pct_counts_mt\", \"pct_counts_rb\", \"pct_counts_hb\"]\n",
    "LOG_TRANSFORMED: list[str] = [\"total_counts\", \"n_genes_by_counts\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "\n",
    "plotting_qc_keys = pf.get_keys(qc_dict)\n",
    "\n",
    "if len(pf.get_keys(qc_dict)) == 0:\n",
    "    raise ValueError(\"Couldn't get valid qc variables from QC dict.\")\n",
    "\n",
    "filter_qc_keys = plotting_qc_keys\n",
    "\n",
    "plotting_qc_keys = plotting_qc_keys + [\"pct_counts_rb\"]\n",
    "\n",
    "if FILTER_DOUBLETS and \"doublet_score\" not in plotting_qc_keys:\n",
    "    plotting_qc_keys = plotting_qc_keys + [\"doublet_score\"]\n",
    "\n",
    "if CORRECT_AMBIENT_RNA and \"decontX_contamination\" not in plotting_qc_keys:\n",
    "    plotting_qc_keys = plotting_qc_keys + [\"decontX_contamination\"]\n",
    "\n",
    "if CELL_CYCLE_SCORE and \"S_score\" not in plotting_qc_keys:\n",
    "    plotting_qc_keys = plotting_qc_keys + [\"S_score\"]\n",
    "\n",
    "if CELL_CYCLE_SCORE and \"G2M_score\" not in plotting_qc_keys:\n",
    "    plotting_qc_keys += [\"G2M_score\"]\n",
    "\n",
    "\n",
    "sample_cls: type[AutoDiscover] = discover_factory[TECHNOLOGY]\n",
    "sample_utils = sample_cls(DIR_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "\n",
    "# Diagnosic pipeline\n",
    "## Reading files\n",
    "\n",
    "if AUTODISCOVER and len(samples) == 0:\n",
    "    samples: dict[str, str] = sample_utils.get_samples()\n",
    "    read_function: Callable = sample_utils.read_function()\n",
    "else:\n",
    "    samples = config[\"basic\"][\"samples\"]\n",
    "    read_function = sample_utils.read_function(samples)\n",
    "    if len(samples) == 0:\n",
    "        raise RuntimeError(\n",
    "            \"No samples paths were provided, provide sample paths as a dictionary in 'config.toml'\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | warning: false\n",
    "itables.show(pd.DataFrame(samples, index=[\"sample path\"]).T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "\n",
    "adatas = {}\n",
    "for sample_id, filename in samples.items():\n",
    "    sample_adata = read_function(filename)\n",
    "    sample_adata.var_names_make_unique()\n",
    "    adatas[sample_id] = sample_adata\n",
    "\n",
    "adata = ad.concat(adatas, label=\"sample\", join=\"outer\", merge=\"same\")\n",
    "adata.obs_names_make_unique()\n",
    "del adatas\n",
    "\n",
    "if TECHNOLOGY == \"10x\" and CORRECT_AMBIENT_RNA:\n",
    "    adatas_raw = {}\n",
    "    if AUTODISCOVER:\n",
    "        raw_samples: dict[str, str] = sample_utils.get_raw_samples()\n",
    "        raw_read_fun = sample_utils.raw_read_function()\n",
    "    else:\n",
    "        raw_read_fun = sample_utils.raw_read_function(raw_samples)\n",
    "\n",
    "    for sample_id, filename in raw_samples.items():  # type: ignore\n",
    "        raw = raw_read_fun(filename)\n",
    "        raw.var_names_make_unique()\n",
    "        adatas_raw[sample_id] = raw\n",
    "\n",
    "    adata_raw = ad.concat(adatas_raw, label=\"sample\", join=\"outer\", merge=\"same\")\n",
    "    adata_raw.obs_names_make_unique()\n",
    "    del adatas_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "## Adding quality metrics\n",
    "mt_features = pf.qc_features_rules[ORGANISM][\"mito\"]\n",
    "rb_features = pf.qc_features_rules[ORGANISM][\"ribo\"]\n",
    "hb_features = pf.qc_features_rules[ORGANISM][\"hb\"]\n",
    "\n",
    "# mitochondrial genes, \"MT-\" for human, \"Mt-\" for mouse\n",
    "adata.var[\"mt\"] = adata.var_names.str.startswith(tuple(mt_features))\n",
    "# ribosomal genes\n",
    "adata.var[\"rb\"] = adata.var_names.str.startswith(tuple(rb_features))\n",
    "# hemoglobin genes\n",
    "adata.var[\"hb\"] = adata.var_names.str.contains(\n",
    "    tuple(hb_features)[0]\n",
    ")  # Only regex is accepted\n",
    "\n",
    "sc.pp.calculate_qc_metrics(\n",
    "    adata, qc_vars=[\"mt\", \"rb\", \"hb\"], percent_top=[20], inplace=True, log1p=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "\n",
    "# Validate QC dict\n",
    "if not pf.validate_qc_dict(qc_dict, adata.obs):\n",
    "    raise ValueError(\"Invalid QC dict, check the keys and values in the QC dict.\")\n",
    "\n",
    "# Convert global qc_dict to per-sample qc_dict\n",
    "if QC_PER_SAMPLE:\n",
    "    if all(map(lambda x: isinstance(x, (list, Number)), qc_dict.values())):\n",
    "        qc_dict = {sample: qc_dict for sample in adata.obs[\"sample\"].unique()}\n",
    "\n",
    "if QC_PER_SAMPLE:\n",
    "    outlier_results = []\n",
    "    for sample_id in adata.obs[\"sample\"].unique():\n",
    "        sample_obs: pd.DataFrame = adata.obs.query(\"sample == @sample_id\")\n",
    "        sample_outliers: pd.DataFrame = pf.compute_outliers(\n",
    "            sample_obs, qc_dict[sample_id], MAX_ONLY, LOG_TRANSFORMED\n",
    "        )\n",
    "        outlier_results.append(sample_outliers)\n",
    "    adata.obs = pd.concat(outlier_results)\n",
    "\n",
    "else:\n",
    "    adata.obs = pf.compute_outliers(adata.obs, qc_dict, MAX_ONLY, LOG_TRANSFORMED)\n",
    "\n",
    "\n",
    "# Aggregate QC metrics\n",
    "outlier_keys = [x + \"_outlier\" for x in pf.get_keys(qc_dict)]\n",
    "outlier_keys = [x for x in outlier_keys if x in adata.obs.columns]\n",
    "adata.obs[\"outliers\"] = adata.obs[outlier_keys].any(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "\n",
    "## Ambient RNA correction\n",
    "# TODO: Check if the Ambient RNA can be improved by using Batch information?\n",
    "\n",
    "if CORRECT_AMBIENT_RNA and TECHNOLOGY == \"10x\":\n",
    "    import tempfile\n",
    "    from os import path, system\n",
    "\n",
    "    with tempfile.TemporaryDirectory(dir=\".\") as tmpdirname:\n",
    "\n",
    "        # Workaround failure of GenomeInfoDb to install in Pixi\n",
    "        pf.GenomeInfoDB_fix(tmpdirname)\n",
    "        # Define paths for temporary files\n",
    "        sce_path = path.join(tmpdirname, \"sce.h5ad\")\n",
    "        raw_path = path.join(tmpdirname, \"raw.h5ad\")\n",
    "        decontx_path = path.join(tmpdirname, \"decontX.h5ad\")\n",
    "\n",
    "        # Save adata and adata_raw to the temporary directory\n",
    "        adata.write_h5ad(sce_path)\n",
    "        adata_raw.write_h5ad(raw_path)\n",
    "\n",
    "        # Execute R scripts with temporary file paths\n",
    "        system(\n",
    "            f\"Rscript ./utils/deconx.R -s {sce_path} -r {raw_path} -o {decontx_path}\"\n",
    "        )\n",
    "\n",
    "        # Read the result back from the temporary directory\n",
    "        adata = sc.read_h5ad(decontx_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "## Doublet Detection\n",
    "\n",
    "# TODO: Check real-life performance\n",
    "# TODO: Check Interop with R to convert object to R & vice-versa\n",
    "if FILTER_DOUBLETS and TECHNOLOGY == \"10x\":\n",
    "    sc.pp.scrublet(adata, batch_key=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "## Cell cycle Scoring\n",
    "# **Not reliable, do via interop later**\n",
    "if not \"counts\" in adata.layers.keys():\n",
    "    adata.layers[\"counts\"] = adata.X.copy()\n",
    "else:\n",
    "    adata.X = adata.layers[\"counts\"].copy()\n",
    "\n",
    "sc.pp.normalize_total(adata)\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.scale(adata)\n",
    "\n",
    "\n",
    "if CELL_CYCLE_SCORE:\n",
    "    if ORGANISM in [\"human\", \"mouse\"]:\n",
    "        s_genes = [x.strip() for x in open(path.join(DIR_RESOURCES, \"s_genes.txt\"))]\n",
    "        g2m_genes = [x.strip() for x in open(path.join(DIR_RESOURCES, \"g2m_genes.txt\"))]\n",
    "\n",
    "        if ORGANISM == \"mouse\":\n",
    "            s_genes = pf.human2mouse(s_genes)\n",
    "            g2m_genes = pf.human2mouse(g2m_genes)\n",
    "\n",
    "        s_genes = [x for x in s_genes if x in adata.var_names]\n",
    "        g2m_genes = [x for x in g2m_genes if x in adata.var_names]\n",
    "\n",
    "        # Cell cycle scoring is not reliable and not similair to Seurat\n",
    "        sc.tl.score_genes_cell_cycle(adata=adata, s_genes=s_genes, g2m_genes=g2m_genes)\n",
    "    else:\n",
    "        logging.error(\"Organism must be either human or mouse.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Diagnostic Plots (prior to filtering)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic QC plots & metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = pf.create_panel_fig(total_plots=len(plotting_qc_keys))\n",
    "\n",
    "# Prevent the subplots from showing\n",
    "plt.close(fig)\n",
    "\n",
    "for i, key in enumerate(plotting_qc_keys):\n",
    "    sc.pl.violin(\n",
    "        adata, keys=[key], groupby=\"sample\", stripplot=False, inner=\"box\", ax=axs[i]\n",
    "    )\n",
    "    axs[i].set_xticklabels(axs[i].get_xticklabels(), rotation=30)\n",
    "\n",
    "plt.tight_layout()\n",
    "display(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of basic QC metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1 = adata.obs.groupby(\"sample\")[plotting_qc_keys].agg([\"mean\", \"median\"]).round(3)\n",
    "df2 = adata.obs.groupby(\"sample\")[[\"sample\"]].agg([\"size\"])\n",
    "itables.show(pd.concat([df1, df2], axis=1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Histograms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Interactive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "\n",
    "sample_fig_list = []\n",
    "hvplot.extension(\"bokeh\")\n",
    "figs = []\n",
    "df = adata.obs[plotting_qc_keys]\n",
    "for key in df.columns:\n",
    "    fig = (\n",
    "        adata.obs[[key, \"sample\"]]\n",
    "        .hvplot(\n",
    "            kind=\"hist\",\n",
    "            bins=300,\n",
    "            width=400,\n",
    "            height=350,\n",
    "            by=\"sample\",\n",
    "            line_color=None,\n",
    "            bgcolor=\"white\",\n",
    "            shared_axes=False,\n",
    "        )\n",
    "        .opts(axiswise=True)\n",
    "    )\n",
    "    figs.append(fig)\n",
    "\n",
    "layout = reduce(add, figs)\n",
    "layout = layout.cols(2)\n",
    "sample_fig_list.append((\"All samples\", layout))\n",
    "\n",
    "samples = adata.obs[\"sample\"].unique()\n",
    "samples = np.sort(samples)\n",
    "for sample in samples:\n",
    "    figs = []\n",
    "    df = adata.obs.loc[adata.obs[\"sample\"] == sample, plotting_qc_keys]\n",
    "    for key in df.columns:\n",
    "        fig = (\n",
    "            df[[key]]\n",
    "            .hvplot(\n",
    "                kind=\"hist\",\n",
    "                bins=300,\n",
    "                width=400,\n",
    "                height=350,\n",
    "                line_color=None,\n",
    "                bgcolor=\"white\",\n",
    "                shared_axes=False,\n",
    "            )\n",
    "            .opts(axiswise=True)\n",
    "        )\n",
    "        figs.append(fig)\n",
    "    layout = reduce(add, figs)\n",
    "    layout = layout.cols(2)\n",
    "    sample_fig_list.append((sample, layout))\n",
    "\n",
    "\n",
    "pn.Tabs(*sample_fig_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Static"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figs = []\n",
    "\n",
    "ls = []\n",
    "# Plot a histogram on each subplot\n",
    "for i, col in enumerate(plotting_qc_keys):\n",
    "    with rc_context(rc={\"figure.figsize\": (4, 3)}):\n",
    "        fig, ax = plt.subplots()\n",
    "        sns.histplot(data=adata.obs, x=col, hue=\"sample\", bins=300, ax=ax)\n",
    "        sns.move_legend(ax, \"best\", bbox_to_anchor=(1, 1))\n",
    "        plt.setp(ax.get_legend().get_texts(), fontsize=\"10\")  # for legend text\n",
    "        plt.setp(ax.get_legend().get_title(), fontsize=\"10\")  # for legend title\n",
    "        plt.tight_layout()\n",
    "        ls.append(pn.pane.Matplotlib(fig))\n",
    "        plt.close(fig)\n",
    "\n",
    "layout = pn.GridBox(*ls, ncols=2)\n",
    "# layout = layout.cols(2)\n",
    "figs.append((\"All samples\", layout))\n",
    "\n",
    "\n",
    "samples = adata.obs[\"sample\"].unique()\n",
    "samples = np.sort(samples)\n",
    "for sample in samples:\n",
    "    df = adata.obs.loc[adata.obs[\"sample\"] == sample, plotting_qc_keys]\n",
    "    fig, axs = pf.create_panel_fig(total_plots=len(plotting_qc_keys))\n",
    "\n",
    "    # Plot a histogram on each subplot\n",
    "    for i, col in enumerate(plotting_qc_keys):\n",
    "        df[col].plot(kind=\"hist\", ax=axs[i], bins=300)\n",
    "    plt.tight_layout()\n",
    "    figs.append((sample, pn.pane.Matplotlib(fig)))\n",
    "    plt.close(fig)\n",
    "\n",
    "pn.Tabs(*figs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scatter plots of confounders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "\n",
    "df = adata.obs[plotting_qc_keys + [\"sample\"]]\n",
    "fig, axs = pf.create_panel_fig(total_plots=len(plotting_qc_keys), figsize=4)\n",
    "\n",
    "\n",
    "sns.scatterplot(\n",
    "    df, x=\"total_counts\", y=\"n_genes_by_counts\", hue=\"sample\", alpha=0.4, s=6, ax=axs[0]\n",
    ")\n",
    "axs[0].legend(fancybox=True, framealpha=0.8)\n",
    "sns.scatterplot(\n",
    "    df, x=\"total_counts\", y=\"pct_counts_mt\", hue=\"sample\", alpha=0.4, s=6, ax=axs[1]\n",
    ")\n",
    "axs[1].legend(fancybox=True, framealpha=0.8)\n",
    "sns.scatterplot(\n",
    "    df, x=\"total_counts\", y=\"pct_counts_rb\", hue=\"sample\", alpha=0.4, s=6, ax=axs[2]\n",
    ")\n",
    "axs[2].legend(fancybox=True, framealpha=0.8)\n",
    "sns.move_legend(axs[0], \"upper left\", bbox_to_anchor=(1, 1))\n",
    "sns.move_legend(axs[1], \"upper left\", bbox_to_anchor=(1, 1))\n",
    "sns.move_legend(axs[2], \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "if FILTER_DOUBLETS:\n",
    "    sns.scatterplot(\n",
    "        df, x=\"total_counts\", y=\"doublet_score\", hue=\"sample\", alpha=0.4, s=6, ax=axs[3]\n",
    "    )\n",
    "    sns.move_legend(axs[3], \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "if CORRECT_AMBIENT_RNA:\n",
    "    sns.scatterplot(\n",
    "        df,\n",
    "        x=\"total_counts\",\n",
    "        y=\"decontX_contamination\",\n",
    "        hue=\"sample\",\n",
    "        alpha=0.4,\n",
    "        s=6,\n",
    "        ax=axs[4],\n",
    "    )\n",
    "    sns.move_legend(axs[4], \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "for i, ax in enumerate(axs):\n",
    "    if not ax.has_data():\n",
    "        ax.set_visible(False)\n",
    "\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DeconX contamination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "\n",
    "# TODO: Test this plot\n",
    "if CORRECT_AMBIENT_RNA and TECHNOLOGY == \"10x\":\n",
    "    fig, axs = pf.create_panel_fig(total_plots=2, figsize=4)\n",
    "    scatter = sns.scatterplot(\n",
    "        pd.concat([adata.obsm[\"decontX_UMAP\"], adata.obs], axis=1),\n",
    "        x=\"DecontX_UMAP_1\",\n",
    "        y=\"DecontX_UMAP_2\",\n",
    "        hue=\"decontX_contamination\",\n",
    "        s=0.8,\n",
    "        ax=axs[0],\n",
    "        palette=\"inferno\",\n",
    "    )\n",
    "    sns.move_legend(\n",
    "        scatter, \"center right\", bbox_to_anchor=(1.2, 0.5), title=None, frameon=False\n",
    "    )\n",
    "\n",
    "    scatter = sns.scatterplot(\n",
    "        pd.concat([adata.obsm[\"decontX_UMAP\"], adata.obs], axis=1),\n",
    "        x=\"DecontX_UMAP_1\",\n",
    "        y=\"DecontX_UMAP_2\",\n",
    "        hue=\"decontX_clusters\",\n",
    "        s=0.8,\n",
    "        ax=axs[1],\n",
    "    )\n",
    "    sns.move_legend(\n",
    "        scatter, \"center right\", bbox_to_anchor=(1.2, 0.5), title=None, frameon=False\n",
    "    )\n",
    "\n",
    "    plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Effects of confounders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "sc.pp.pca(adata, n_comps=30)\n",
    "sc.pp.neighbors(adata)\n",
    "sc.tl.leiden(adata, key_added=\"groups\", flavor=\"leidenalg\")\n",
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "figs = sc.pl.umap(\n",
    "    adata,\n",
    "    size=7,\n",
    "    color=[\"sample\"] + plotting_qc_keys,\n",
    "    show=False,\n",
    "    ncols=2,\n",
    "    color_map=\"inferno\",\n",
    "    sort_order=False,\n",
    "    alpha=0.8,\n",
    "    hspace=0.2,\n",
    "    wspace=0.3,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Filtering "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filtering thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "\n",
    "ls = {}\n",
    "for key in filter_qc_keys:\n",
    "    ls[key] = (\n",
    "        adata.obs.query(f\"{key}_outlier == False\").groupby(\"sample\")[key].agg(\"min\")\n",
    "    )\n",
    "min_df = pd.concat(ls, axis=1)\n",
    "\n",
    "for key in filter_qc_keys:\n",
    "    ls[key] = (\n",
    "        adata.obs.query(f\"{key}_outlier == False\").groupby(\"sample\")[key].agg(\"max\")\n",
    "    )\n",
    "max_df = pd.concat(ls, axis=1)\n",
    "\n",
    "pd.concat([min_df, max_df], keys=[\"min\", \"max\"], axis=0).T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Number of outliers based on provided criteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "\n",
    "df_l = []\n",
    "col_l = []\n",
    "for key in filter_qc_keys:\n",
    "    try:\n",
    "        df_l.append(\n",
    "            pd.DataFrame(\n",
    "                adata.obs[[f\"{key}_outlier\", \"sample\"]].value_counts().loc[(True), :]\n",
    "            )\n",
    "        )\n",
    "        col_l.append(f\"{key}_outlier\")\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "col_l.append(\"Aggreagte Outliers\")\n",
    "\n",
    "\n",
    "df_l.append(\n",
    "    pd.DataFrame(adata.obs[[\"outliers\", \"sample\"]].value_counts().loc[(True), :])\n",
    ")\n",
    "\n",
    "df_l.append(pd.DataFrame(adata.obs[[\"sample\"]].value_counts()))\n",
    "df_l[-1].index = df_l[-1].index.get_level_values(0)\n",
    "col_l.append(\"Total Cells\")\n",
    "\n",
    "df = pd.concat(df_l, axis=1)\n",
    "df.columns = [col_l]\n",
    "df = df.loc[np.sort(df.index), :]\n",
    "itables.show(df.fillna(0))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell filtering based on outlier function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "adata.obs = adata.obs.infer_objects()\n",
    "\n",
    "# Saving The object at the last step before subsseting\n",
    "if path.exists(DIR_SAVE):\n",
    "    adata.write_h5ad(path.join(DIR_SAVE, \"raw_adata.h5ad\"))\n",
    "else:\n",
    "    mkdir(DIR_SAVE)\n",
    "    adata.write_h5ad(path.join(DIR_SAVE, \"raw_adata.h5ad\"))\n",
    "\n",
    "# Cell Filtering based on threshold\n",
    "\n",
    "\n",
    "adata = adata[(~adata.obs.outliers)].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# QC plots (post filtering)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table of basic QC metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "\n",
    "df1 = adata.obs.groupby(\"sample\")[plotting_qc_keys].agg([\"mean\", \"median\"]).round(3)\n",
    "df2 = adata.obs.groupby(\"sample\")[[\"sample\"]].agg([\"size\"])\n",
    "df = pd.concat([df1, df2], axis=1)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Histograms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Interactive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "\n",
    "\n",
    "from operator import add\n",
    "\n",
    "sample_fig_list = []\n",
    "hvplot.extension(\"bokeh\")\n",
    "ls = []\n",
    "df = adata.obs[plotting_qc_keys]\n",
    "for key in df.columns:\n",
    "    fig = (\n",
    "        adata.obs[[key, \"sample\"]]\n",
    "        .hvplot(\n",
    "            kind=\"hist\",\n",
    "            bins=300,\n",
    "            width=400,\n",
    "            height=350,\n",
    "            by=\"sample\",\n",
    "            line_color=None,\n",
    "            bgcolor=\"white\",\n",
    "            shared_axes=False,\n",
    "        )\n",
    "        .opts(axiswise=True)\n",
    "    )\n",
    "    ls.append(fig)\n",
    "\n",
    "layout = reduce(add, ls)\n",
    "layout = layout.cols(2)\n",
    "sample_fig_list.append((\"All samples\", layout))\n",
    "\n",
    "samples = adata.obs[\"sample\"].unique()\n",
    "samples = np.sort(samples)\n",
    "for sample in samples:\n",
    "    ls = []\n",
    "    df = adata.obs.loc[adata.obs[\"sample\"] == sample, plotting_qc_keys]\n",
    "    for key in df.columns:\n",
    "        fig = (\n",
    "            df[[key]]\n",
    "            .hvplot(\n",
    "                kind=\"hist\",\n",
    "                bins=300,\n",
    "                width=400,\n",
    "                height=350,\n",
    "                line_color=None,\n",
    "                bgcolor=\"white\",\n",
    "                shared_axes=False,\n",
    "            )\n",
    "            .opts(axiswise=True)\n",
    "        )\n",
    "        ls.append(fig)\n",
    "    layout = reduce(add, ls)\n",
    "    layout = layout.cols(2)\n",
    "    sample_fig_list.append((sample, layout))\n",
    "\n",
    "\n",
    "pn.Tabs(*sample_fig_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Static"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figs = []\n",
    "\n",
    "ls = []\n",
    "# Plot a histogram on each subplot\n",
    "for i, col in enumerate(plotting_qc_keys):\n",
    "    with rc_context(rc={\"figure.figsize\": (4, 3)}):\n",
    "        fig, ax = plt.subplots()\n",
    "        sns.histplot(data=adata.obs, x=col, hue=\"sample\", bins=300, ax=ax)\n",
    "        sns.move_legend(ax, \"best\", bbox_to_anchor=(1, 1))\n",
    "        plt.setp(ax.get_legend().get_texts(), fontsize=\"10\")  # for legend text\n",
    "        plt.setp(ax.get_legend().get_title(), fontsize=\"10\")  # for legend title\n",
    "        plt.tight_layout()\n",
    "        ls.append(pn.pane.Matplotlib(fig))\n",
    "        plt.close(fig)\n",
    "\n",
    "layout = pn.GridBox(*ls, ncols=2)\n",
    "# layout = layout.cols(2)\n",
    "figs.append((\"All samples\", layout))\n",
    "\n",
    "\n",
    "# figs.append((\"All Samples\", pn.pane.Matplotlib(fig)))\n",
    "\n",
    "\n",
    "samples = adata.obs[\"sample\"].unique()\n",
    "samples = np.sort(samples)\n",
    "for sample in samples:\n",
    "    df = adata.obs.loc[adata.obs[\"sample\"] == sample, plotting_qc_keys]\n",
    "    fig, axs = pf.create_panel_fig(total_plots=len(plotting_qc_keys))\n",
    "\n",
    "    # Plot a histogram on each subplot\n",
    "    for i, col in enumerate(plotting_qc_keys):\n",
    "        df[col].plot(kind=\"hist\", ax=axs[i], bins=300)\n",
    "    plt.tight_layout()\n",
    "    figs.append((sample, pn.pane.Matplotlib(fig)))\n",
    "    plt.close(fig)\n",
    "\n",
    "pn.Tabs(*figs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Scatter plots of confounders after filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "\n",
    "# Assuming 'df' is your DataFrame\n",
    "df = adata.obs[plotting_qc_keys + [\"sample\"]]\n",
    "fig, axs = pf.create_panel_fig(total_plots=len(df.columns), figsize=4)\n",
    "\n",
    "sns.scatterplot(\n",
    "    df, x=\"total_counts\", y=\"n_genes_by_counts\", hue=\"sample\", alpha=0.4, s=6, ax=axs[0]\n",
    ")\n",
    "axs[0].legend(fancybox=True, framealpha=0.8)\n",
    "sns.scatterplot(\n",
    "    df, x=\"total_counts\", y=\"pct_counts_mt\", hue=\"sample\", alpha=0.4, s=6, ax=axs[1]\n",
    ")\n",
    "axs[1].legend(fancybox=True, framealpha=0.8)\n",
    "sns.scatterplot(\n",
    "    df, x=\"total_counts\", y=\"pct_counts_rb\", hue=\"sample\", alpha=0.4, s=6, ax=axs[2]\n",
    ")\n",
    "axs[2].legend(fancybox=True, framealpha=0.8)\n",
    "sns.move_legend(axs[0], \"upper left\", bbox_to_anchor=(1, 1))\n",
    "sns.move_legend(axs[1], \"upper left\", bbox_to_anchor=(1, 1))\n",
    "sns.move_legend(axs[2], \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "if FILTER_DOUBLETS:\n",
    "    sns.scatterplot(\n",
    "        df, x=\"total_counts\", y=\"doublet_score\", hue=\"sample\", alpha=0.4, s=6, ax=axs[3]\n",
    "    )\n",
    "    sns.move_legend(axs[3], \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "if CORRECT_AMBIENT_RNA:\n",
    "    sns.scatterplot(\n",
    "        df,\n",
    "        x=\"total_counts\",\n",
    "        y=\"decontX_contamination\",\n",
    "        hue=\"sample\",\n",
    "        alpha=0.4,\n",
    "        s=6,\n",
    "        ax=axs[4],\n",
    "    )\n",
    "    sns.move_legend(axs[4], \"upper left\", bbox_to_anchor=(1, 1))\n",
    "\n",
    "[ax.set_visible(False) for ax in axs.flatten() if not ax.has_data()]\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DecontX contamination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "\n",
    "if CORRECT_AMBIENT_RNA and TECHNOLOGY == \"10x\":\n",
    "    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 5))\n",
    "\n",
    "    scatter = sns.scatterplot(\n",
    "        pd.concat([adata.obsm[\"decontX_UMAP\"], adata.obs], axis=1),\n",
    "        x=\"DecontX_UMAP_1\",\n",
    "        y=\"DecontX_UMAP_2\",\n",
    "        hue=\"decontX_contamination\",\n",
    "        s=0.8,\n",
    "        ax=ax1,\n",
    "        palette=\"inferno\",\n",
    "    )\n",
    "    sns.move_legend(\n",
    "        scatter, \"center right\", bbox_to_anchor=(1.1, 0.5), title=None, frameon=False\n",
    "    )\n",
    "\n",
    "    scatter = sns.scatterplot(\n",
    "        pd.concat([adata.obsm[\"decontX_UMAP\"], adata.obs], axis=1),\n",
    "        x=\"DecontX_UMAP_1\",\n",
    "        y=\"DecontX_UMAP_2\",\n",
    "        hue=\"decontX_clusters\",\n",
    "        s=0.8,\n",
    "        ax=ax2,\n",
    "    )\n",
    "    sns.move_legend(\n",
    "        scatter, \"center right\", bbox_to_anchor=(1.15, 0.5), title=None, frameon=False\n",
    "    )\n",
    "\n",
    "    plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if CORRECT_AMBIENT_RNA and TECHNOLOGY == \"10x\":\n",
    "    with rc_context(\n",
    "        {\n",
    "            \"figure.figsize\": (\n",
    "                adata.obs[\"decontX_clusters\"].astype(\"int32\").max() * 0.7,\n",
    "                5,\n",
    "            )\n",
    "        }\n",
    "    ):\n",
    "        sns.violinplot(adata.obs, x=\"decontX_clusters\", y=\"decontX_contamination\")\n",
    "        sns.stripplot(\n",
    "            adata.obs, x=\"decontX_clusters\", y=\"decontX_contamination\", s=1, c=\"black\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clustering after to cell filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "sc.pp.pca(adata, n_comps=30)\n",
    "sc.pp.neighbors(adata)\n",
    "sc.tl.leiden(adata, key_added=\"groups\", flavor=\"leidenalg\")\n",
    "sc.tl.umap(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | warning: false\n",
    "with rc_context({\"figure.figsize\": (5, (len(plotting_qc_keys) + 1) * 0.6)}):\n",
    "    figs = sc.pl.umap(\n",
    "        adata,\n",
    "        size=8,\n",
    "        color=[\"sample\"] + plotting_qc_keys,\n",
    "        show=False,\n",
    "        ncols=2,\n",
    "        color_map=\"inferno\",\n",
    "        sort_order=False,\n",
    "        alpha=0.8,\n",
    "        hspace=0.2,\n",
    "        wspace=0.3,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# | echo: false\n",
    "# | output: false\n",
    "# | warning: false\n",
    "\n",
    "adata.obs = adata.obs.infer_objects()\n",
    "## Save Result\n",
    "if path.exists(DIR_SAVE):\n",
    "    adata.write_h5ad(path.join(DIR_SAVE, \"adata.h5ad\"))\n",
    "else:\n",
    "    mkdir(DIR_SAVE)\n",
    "    adata.write_h5ad(path.join(DIR_SAVE, \"adata.h5ad\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Session Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "session_info.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Pixi - Python 3 (ipykernel)",
   "language": "python",
   "name": "pixi-kernel-python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
